### PR TITLE
Set informational Network Policy validation condition in the operator.

### PR DESCRIPTION
**Pull request by squeed**
Without this, the only way to determine if a policy is valid is to read the agent logs. This is unfortunate (and impossible for most cluster end-users).

Two primary changes:
- Add an idiomatic Conditions array to CNP and CCNP.   
- Cell to the operator that watches all CNPs and CCNPs. It validates policies using the same logic as the agent, setting an informational Condition when a policy is not valid.

```release-note
CiliumNetworkPolicies are now validated by the operator and the result set in the object's Status field.
```

### CODE REVIEW

File: operator/pkg/networkpolicy/cell.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	k8s_client "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var Cell = cell.Module(
+	"network-policy-validator",
+	"Validates network policies and reports their validiity status",
```
Comment (pippolo84) [inline operator/pkg/networkpolicy/cell.go]: Typo: `validiity` -> `validity`

**Review by pippolo84 — APPROVED** [review]
Nice, thanks! :100:

**Review by nebril — APPROVED** [review]
contributing LGTM

**Review by learnitall — APPROVED** [review]
API changes LGTM, thanks!

File: operator/pkg/networkpolicy/cell.go (hunk 1 above)
Comment (christarazi) [inline operator/pkg/networkpolicy/cell.go]: Maybe we should clarify that we are only doing validation of CNP & CCNP, and not KNPs.

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go:1]: Because we are changing the CRD, we need to bump `CustomResourceDefinitionSchemaVersion`

**Review by christarazi — COMMENTED** [review]
Nice improvement! Just a few small comments to address

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (squeed) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go:1]: Do we? I thought we only had to bump it between released versions?

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go [hunk 2]
```diff
@@ -77,6 +77,16 @@ type CiliumNetworkPolicyStatus struct {
 	// DerivativePolicies is the status of all policies derived from the Cilium
 	// policy
 	DerivativePolicies map[string]CiliumNetworkPolicyNodeStatus `json:"derivativePolicies,omitempty"`
+
+	// The generation observed by the policy controller
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
Comment (squeed) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go]: In retrospect, this is not a useful field; I'll remove it.

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go:1]: It was bumped as well in https://github.com/cilium/cilium/pull/32382. If we didn't bump it, devs deploying various points of `main` would not get their CRDs updated. AFAIK, when it comes to the release, we "normalize/squash" the bumped version.

Comment (squeed) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go:1]: You're right, somehow I had the wrong impression. Fixed.

**Review by christarazi — APPROVED** [review]

File: operator/pkg/networkpolicy/cell.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		log.WithError(err).Error("failed to update CCNP status")
+	}
+
+	return err
+}
+
+// updateCondition creates or updates the policy validation condition in Conditions, setting
+// the transition time as necessary.
+func updateCondition(conditions []cilium_api_v2.NetworkPolicyCondition, errs error) []cilium_api_v2.NetworkPolicyCondition {
+	wantCondition := corev1.ConditionFalse
```
Comment (gandro) [inline operator/pkg/networkpolicy/cell.go]: Shouldn't this be `ConditionTrue` to clear the condition once errs is nil?

**Review by gandro — COMMENTED** [review]
Apologies for the late review. Really nice idea, I like it! One question in the code

File: operator/pkg/networkpolicy/cell.go (hunk 3 above)
Comment (squeed) [reply operator/pkg/networkpolicy/cell.go]: argh! good catch.

**Review by gandro — APPROVED** [review]

**Pull request merged**